### ISSUE TITLE
Proposal: "rollback catalog" database recovery operation

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#3381]
(Companion to https://github.com/influxdata/influxdb_iox/issues/3380)

__Use case:__
IOx has been ingesting fine for a large period of time, writing data to the object store; Then there is some problem with ingestion that causes it to persist "bad" data for some period of time. For example:

1. A software bug that writes bad stuff to object store
2. Some sort of kafka problem that results in data loss
3. Operator error  (e.g. overzealous use of the feature listed in https://github.com/influxdata/influxdb_iox/issues/3380)

Now, we would like to get IOx back to a *consistent* state, where it has ingested all the data once, in a reasonable amount of time (really As Soon As Possible!!!)

Today the options are limited:
1. Wipe the catalog, and re-ingest all data from the beginning of time from the out of band backup (likely to be a prohibitively lengthy process for large databases)
2. Reingest the data from some sort of out of band backup (and rely on the native deduplication process of IOx to converge to a consistent state)

__Proposal:__

@mkmik @crepererum and I propose a `rollback-catalog` primitive to the `recover` database command:

Example usage:
```shell
# Roll catalog back to a specific transaction id
./target/debug/influxdb_iox database recover rollback-catalog <db_name> --transaction_id=XX

# Roll catalog back back to the transaction nearest to a specific UTC time: "2d ago", "2021-01-02T10:13:32", etc)
./target/debug/influxdb_iox database recover rollback-catalog <db_name> --time=2021-01-02T10:13:32
```

This would instruct the database, in whatever initialization state it happened to be to:
1. stop accepting new writes momentarily
2. discard all data currently in memory
3. Discard all catalog transactions that occurred after the specified one
4. Reinitialize as though it had started afresh

The command would error if
1. The specified time period did not have any persisted transactions
2. The rollback period was prior to the oldest snapshot (as the catalog does not have individual transactions prior to that point)

It is also possible that the database would fail to restart successfully after catalog rollback because the data needed for replay would no longer be present in the sequencer. In this case, a feature such as https://github.com/influxdata/influxdb_iox/issues/3380 would be handy to get the data flowing again

__Implementation Thoughts__
@crepererum  thinks the existing software architecture could support this feature, but it needs some work at the edges (e.g. CLI, testing, etc) to implement it

__Open Questions__
1. Would there be some way to "undo" a rollback operation?

**Comment by NGA-TRAN** [comment]
@alamb and @crepererum : Can you clarify step 4 `Reinitialize as though it had started afresh`? Does it  mean we rebuild the catalog from the latest DB checkpoint recorded in the preserved catalog. Then after the catalog is rebuilt, we will accept writes starting from the sequence number after that checkpoint?

**Comment by mkmik** [comment]
> Would there be some way to "undo" a rollback operation?

yeah it would be nice if our catalog implemented the model of git.

Basically this means that the old catalog "head" stays around and can be used to later if we changed our minds.
This opens a whole new problem of how we "prune" these catalog "branches".

I wonder why we can't just use git; IIRC back in the days there was some tools that used S3 as direct storage for a git repo and supported pushes and pulls.

**Comment by crepererum** [comment]
> Would there be some way to "undo" a rollback operation?

Unlikely, at least w/ the current data structure. The current catalog is a linear transaction stream, so to rollback and start writing again we need to through away transactions that come after the rollback point. We could extend the catalog to allow forking, but I'm not sure it's worth the hassle and complexity.

**Comment by crepererum** [comment]
> > Would there be some way to "undo" a rollback operation?
> 
> yeah it would be nice if our catalog implemented the model of git.
> 
> Basically this means that the old catalog "head" stays around and can be used to later if we changed our minds. This opens a whole new problem of how we "prune" these catalog "branches".
> 
> I wonder why we can't just use git; IIRC back in the days there was some tools that used S3 as direct storage for a git repo and supported pushes and pulls.

In theory you could use GIT, but:

1. there's no pure Rust implementation for it
2. GIT is actually quite complex with all its pack and index files
3. Tools that map GIT to S3 are very likely quite hackish, because checkouts are heavily based on hardlinks

**Comment by alamb** [comment]
I agree with the assessment the idea of trying to use git directly is a good one in theory but very hard to realize in practice

**Comment by alamb** [comment]
I also think this one is urgent @jacobmarble  -- b/c there are 🐛  (🐉  more likely) in the operations of the persistent catalog that we have been working around by nuking the catalog. The sooner we start to fix them the better I think
